src/boot/efi/pe.c

```diff
@@ -224,6 +239,72 @@ static void pe_locate_sections(
                 }
 }
 
+static bool pe_use_this_dtb(
+                const PeSectionHeader section_table[],
+                size_t n_section_table,
+                const void *base,
+                size_t dtb_offset,
+                size_t dtb_size,
+                size_t section_nb) {
+        assert(section_table);
+        assert(n_section_table > 0);
+        assert(base);
+
+        EFI_STATUS err;
+
+        static const void *cached_base = NULL;
+        static const Device *cached_device = NULL;
+
+        const void *dtb = (const uint8_t *) base + dtb_offset;
+
+        err = devicetree_match(dtb, dtb_size);
+        if (err == EFI_SUCCESS)
+                return true;
+        if (err != EFI_UNSUPPORTED)
+                return false;
+
+        /* Firmware does not provide the devicetree, so try matching against a list from .hwids section */
+        const char *compatible = NULL;
+        if (cached_base != base) {
+                cached_base = base;
+                cached_device = NULL;
+
+                static const char *const hwids_section_name[] = { ".hwids", NULL };
+                PeSectionVector hwids_section = {};
+                pe_locate_sections(
+                                section_table,
+                                n_section_table,
+                                hwids_section_name,
+                                PTR_TO_SIZE(base),
+                                &hwids_section);
```

Comment:
this is of course a recursive call to pe_locate_sections. Which I find a bit weird, but should be permissible, but really should come with a comment, because it is a bit weird. i.e. mention this is recursive and why this is not a problem

Reply:
There is no need to cache hwids section because matching result is already cached (`cached_device`). And it won't be O(n²): this search happens only for the first .dtbauto with the same `validate_base`.

Should I still make the suggested changes?

Reply:
hmm, so the static thing i am not fond of either. i mean this is not going to be multithreaded, but I still think without the static thing it would be a lot nicer.

hence, yes please rework this as suggested, and then you could as well pass the "const Device" into the pe_locate_sections_internal() func, instead of a pointer to the hwids section.


```c
static void pe_locate_sections_internal(
                const PeSectionHeader section_table[],
                size_t n_section_table,
                const char *const section_names[],
                size_t validate_base,
                const Device *device, 
                PeSectionVector sections[]) {
        …
}

static bool looking_for_dbauto(const char *const section_names[]) {
        assert(section_names);

        for (size_t i = 0; section_names[i]; i++) 
                if (streq(section_names, ".dbauto"))
                        return true;
         return false;
}

static void pe_locate_sections(
                const PeSectionHeader section_table[],
                size_t n_section_table,
                const char *const section_names[],
                size_t validate_base,
                PeSectionVector sections[]) {

        if (!looking_for_dbauto(section_names))
                return pe_locate_sections_internal(
                                  section_table,
                                  n_section_table,
                                  section_names,
                                  validate_base,
                                  /* device= */ NULL,
                                  sections);

        /* find device ..*/

        pe_locate_sections_internal(…,  (const char*const*) { ".hwids", NULL) …)

        return pe_locate_sections_internal(…);
}
```

yu get the idea

Reply:
Done